File: src/packages/control-ui/src/api/mesh_socks5_bridge.ts
// Control UI TypeScript: Mesh SOCKS5 Exit Bridge API

export type Socks5RouteDisposition = 'direct' | 'mesh_exit' | 'blocked';

export interface MeshExitNode {
  nodeId: string;
  virtualIp: string;
  active: boolean;
}

export class MeshSocks5Bridge {
  public listenPort: number;
  public exitNode: MeshExitNode | null = null;
  public credentials: Map<string, string> = new Map();

  constructor(listenPort: number) {
    this.listenPort = listenPort;
  }

  public setExitNode(node: MeshExitNode): void {
    this.exitNode = node;
  }

  public addUser(user: string, pass: string): void {
    this.credentials.set(user, pass);
  }

  public authenticate(user: string, pass: string): boolean {
    if (this.credentials.size === 0) return true;
    return this.credentials.get(user) === pass;
  }

  public evaluateRoute(
    host: string,
    port: number,
  ): { disposition: Socks5RouteDisposition; targetIp: string | null } {
    if (port <= 0 || port > 65535) return { disposition: 'blocked', targetIp: null };
    if (host === 'localhost' || host === '127.0.0.1')
      return { disposition: 'direct', targetIp: null };

    if (this.exitNode && this.exitNode.active) {
      return { disposition: 'mesh_exit', targetIp: this.exitNode.virtualIp };
    }
    return { disposition: 'direct', targetIp: null };
  }

  public parseGreeting(data: Uint8Array): number {
    if (data.length < 2 || data[0] !== 0x05) return 0xff;
    const nmethods = data[1]!;
    if (data.length < 2 + nmethods) return 0xff;

    const methods = Array.from(data.slice(2, 2 + nmethods));
    if (this.credentials.size === 0) {
      return methods.includes(0x00) ? 0x00 : 0xff;
    }
    return methods.includes(0x02) ? 0x02 : 0xff;
  }

  public craftReply(repCode: number, bindIp: string, bindPort: number): Uint8Array {
    const reply = [0x05, repCode, 0x00, 0x01];
    const octets = bindIp.split('.').map((o) => parseInt(o, 10));
    reply.push(...octets);
    reply.push((bindPort >> 8) & 0xff, bindPort & 0xff);
    return new Uint8Array(reply);
  }
}

File: labs/mobile/donor-archive/tunnel/MeshSocks5Bridge.kt
package com.luminet.android.tunnel

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.util.concurrent.ConcurrentHashMap

enum class AndroidSocks5Disposition {
    DIRECT,
    MESH_EXIT,
    BLOCKED
}

data class AndroidMeshExitNode(
    val nodeId: String,
    val virtualIp: String,
    val active: Boolean
)

class MeshSocks5Bridge(val listenPort: Int) {
    private var exitNode: AndroidMeshExitNode? = null
    private val credentials = ConcurrentHashMap<String, String>()

    fun setExitNode(node: AndroidMeshExitNode) {
        exitNode = node
    }

    fun addUser(user: String, pass: String) {
        credentials[user] = pass
    }

    fun authenticate(user: String, pass: String): Boolean {
        if (credentials.isEmpty()) return true
        return credentials[user] == pass
    }

    fun evaluateRoute(host: String, port: Int): Pair<AndroidSocks5Disposition, String?> {
        if (port <= 0 || port > 65535) return Pair(AndroidSocks5Disposition.BLOCKED, null)
        if (host == "localhost" || host == "127.0.0.1") return Pair(AndroidSocks5Disposition.DIRECT, null)

        val exit = exitNode
        if (exit != null && exit.active) {
            return Pair(AndroidSocks5Disposition.MESH_EXIT, exit.virtualIp)
        }
        return Pair(AndroidSocks5Disposition.DIRECT, null)
    }

    fun parseGreeting(data: ByteArray): Byte {
        if (data.size < 2 || data[0] != 0x05.toByte()) return 0xFF.toByte()
        val nmethods = data[1].toInt() and 0xFF
        if (data.size < 2 + nmethods) return 0xFF.toByte()

        val methods = data.copyOfRange(2, 2 + nmethods)
        return if (credentials.isEmpty()) {
            if (methods.contains(0x00.toByte())) 0x00.toByte() else 0xFF.toByte()
        } else {
            if (methods.contains(0x02.toByte())) 0x02.toByte() else 0xFF.toByte()
        }
    }

    fun craftReply(repCode: Byte, bindIp: String, bindPort: Int): ByteArray {
        val out = ByteArrayOutputStream()
        out.write(0x05)
        out.write(repCode.toInt())
        out.write(0x00) // Reserved
        out.write(0x01) // IPv4

        val ipParts = bindIp.split(".").map { it.toInt().toByte() }
        out.write(ipParts.toByteArray())

        val portBuf = ByteBuffer.allocate(2)
        portBuf.putShort(bindPort.toShort())
        out.write(portBuf.array())

        return out.toByteArray()
    }
}
